core/src/main/java/fr/sncf/osrd/new_infra/implementation/signaling/modules/BAL3.java

```diff
@@ -0,0 +1,151 @@
+package fr.sncf.osrd.new_infra.implementation.signaling.modules;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import fr.sncf.osrd.new_infra.api.Direction;
+import fr.sncf.osrd.new_infra.api.reservation.DiDetector;
+import fr.sncf.osrd.new_infra.api.reservation.ReservationInfra;
+import fr.sncf.osrd.new_infra.api.reservation.ReservationRoute;
+import fr.sncf.osrd.new_infra.api.signaling.*;
+import fr.sncf.osrd.new_infra.implementation.RJSObjectParsing;
+import fr.sncf.osrd.new_infra_state.api.InfraStateView;
+import fr.sncf.osrd.new_infra_state.api.ReservationRouteState;
+import fr.sncf.osrd.new_infra_state.api.SignalizationStateView;
+import fr.sncf.osrd.railjson.schema.infra.RJSInfra;
+import fr.sncf.osrd.railjson.schema.infra.trackobjects.RJSSignal;
+import fr.sncf.osrd.utils.graph.EdgeDirection;
+import java.awt.*;
+import java.util.*;
+
+/** This module implements the BAL3 signaling system.
+ * It will eventually be moved to an external module */
+public class BAL3 implements SignalingModule {
+
+    private enum Aspect {
+        GREEN,
+        YELLOW,
+        RED
+    }
+
+    public static class BAL3Signal implements Signal<BAL3Signal.BAL3SignalState> {
+
+        final Set<BAL3Signal> signalDependencies = new HashSet<>();
+        final Set<ReservationRoute> routeDependencies = new HashSet<>();
+
+        public static class BAL3SignalState implements SignalState {
+            public final Aspect aspect;
+
+            public BAL3SignalState(Aspect aspect) {
+                this.aspect = aspect;
+            }
+
+            @Override
+            public int getRGBColor() {
+                return switch (aspect) {
+                    case GREEN -> Color.GREEN.getRGB();
+                    case YELLOW -> Color.YELLOW.getRGB();
+                    case RED -> Color.RED.getRGB();
+                };
+            }
+        }
+
+        @Override
+        public BAL3SignalState getInitialState() {
+            return new BAL3SignalState(Aspect.GREEN);
+        }
+
+        @Override
+        public BAL3SignalState processDependencyUpdate(InfraStateView state, SignalizationStateView signalization) {
+            var blockingRouteStates = Set.of(
+                    ReservationRouteState.Summary.OCCUPIED,
+                    ReservationRouteState.Summary.CONFLICT,
+                    ReservationRouteState.Summary.REQUESTED
+            );
+            for (var route : routeDependencies)
+                if (blockingRouteStates.contains(state.getState(route).getSummary()))
+                    return new BAL3SignalState(Aspect.RED);
```

Comment:
I don't think this is right. Here an example where the code above doesn't work.

Suppose a BAL3 signal protecting a simple switch (base side). This signal has two routes on which it depends. When one of it is `RESERVED` the other one is in `CONFLICT`. However, the signal will be `GREEN` or `YELLOW` (you have to check the signal at the end of the current `RESERVED` route .

Reply:
I changed it into something closer to what we had in railscript, there were other issues.

But this part isn't tested yet, it will be a part of https://github.com/DGEXSolutions/osrd/issues/648.